### PR TITLE
feat!: theme font family weight internally

### PR DESCRIPTION
**Pull request by pretzelhammer**
## Describe the problem this PR addresses

text inside maker components could not be themed

## Describe the changes in this PR

now text in maker components can be themed

## Other information

https://square.github.io/maker/lab/theme-font-family-weight-internally/#/ThemeTest

### migration guide
on updating to this version in your code you'll have to
- rename any used instances of `--color-heading` (exported by MTheme) to `--color-title`
- rename any used instance of `--color-text` (exported by MTheme) to `--color-paragraph`
- IF you were using the `label` text variant's default theme, the defaults `fontWeight: '500'`, `element: 'h6'` and `textTransform: 'uppercase'` have been removed (per Design's request, since labels in Maker aren't the same thing as labels in Website, and also labels in Website are getting updated themselves to fit these new standards)

### CODE REVIEW

File: src/components/Select/src/SelectControl.vue
Comment (landondurnan) [inline src/components/Select/src/SelectControl.vue:172]: There's several components that have an explicit style set for the font-weight and font-family that points to the paragraph -- but these styles seem unnecessary. 

When not using the theme component, every component will inherit whatever font style and weight are set at the top level or the browser default.

When using the theme component, every component will also inherit that as the default since those styles are set on the parent.

So I think we can remove these explicit styles that currently pointing to the font-family / weight for the paragraph since without them it should work exactly the same way.

**Review by landondurnan — CHANGES_REQUESTED** [review]

File: src/components/Theme/src/default-theme.js
Comment (pretzelhammer) [reply src/components/Theme/src/default-theme.js:22]: I went with `headline`, `title`, `paragraph`, and `label` to be consistent with [Shimaa's doc](https://coda.io/d/ECOM-Website_dM7hmP2aSPk/Design-Specs_suSqm) and Website. Since the names are arbitrary can we just update the eng design to be consistent with Website? It's much easier to update an eng design than updating all of Website.

Sure, we can update `headline` to have the same defaults as `title` or to even point into `title`.

Hmmm, I've reviewed Maker docs and `label` means something different in Maker than it does in Website. In Website `label`s are functionally used as sub-titles. Hmmmm. We may need 5 default text variants.

| Maker | Website |
|-|-|
`headline` | `headline`
`title` | `title`
`paragraph` | `paragraph`
`label` | `ui-font`
no Maker equivalent | `label` (used as subtitles)

We can add a `subtitle` to Maker or Website can just do it themselves, since user-defined variants are supported in beta.

File: src/components/Select/src/SelectControl.vue
Comment (pretzelhammer) [reply src/components/Select/src/SelectControl.vue:172]: This is true assuming that Maker is being used to build an entire standalone app. This isn't true in the case of Website which is a large codebase which has a lot of Maker and non-Maker code intermixed with custom CSS overrides everywhere in between. It's better to code defensively and add redundant CSS styles in this scenario.

File: src/components/Theme/src/default-theme.js
Comment (landondurnan) [reply src/components/Theme/src/default-theme.js:22]: My preference is to stick to a definition that's becoming more of a standard for more themeable design systems: https://theme-ui.com/theme-spec

While I often will side with naming conventions that are the same as what we present in our UI, Maker is meant to be used beyond one product for one team. Adopting a larger defined standard helps us more easily make decisions and be interoperable in the future (considering tooling, documentation, and extensibility).

The current product `label` usage is likely to change, I'm not even sure if we're sticking with that name -- it just happens to be a bit more generic than "buttons" and better than "UI Font". It could be used for more things like navigation, controls, or subtitles -- but we haven't mapped that out really. We just know we'd like to provide four options where everywhere `label` is used now has been updated to default to `body`.

File: src/components/Select/src/SelectControl.vue
Comment (landondurnan) [reply src/components/Select/src/SelectControl.vue:172]: I understand your perspective, but where does defensive CSS end? Shouldn't we trust consumers to make decisions about their styles without increasing the weight of our package? Wouldn't this be a good opportunity for Website to streamline / remove styles they don't need as well?

File: src/components/Theme/src/default-theme.js
Comment (pretzelhammer) [reply src/components/Theme/src/default-theme.js:22]: I had a chat with Shimaa and she helped answer my questions, summary:
1. "label"s within the editor need to be renamed/remapped to "title"s, which is how they're being used already anyway
2. "label"s within Maker just refer to form labels & button text, and should map to paragraph font family and weight (will do in this PR)
3. "body" can be renamed to "paragraph" so there's more alignment / less confusion between Maker and Website

Comment (pretzelhammer) [reply src/components/Theme/src/default-theme.js:22]: i updated the pr after my convos with Shimaa, she's signed off on the designs/theming and has said it's okay to rename "heading" to "title" and "body" to "paragraph" to be more consistent with Website

File: src/components/Select/src/SelectControl.vue
Comment (pretzelhammer) [reply src/components/Select/src/SelectControl.vue]: i updated the pr to remove the explicit defensive css styles where possible

**Review by landondurnan — CHANGES_REQUESTED** [review]
This PR doesn't follow the consensus we had of following the Theme UI spec for the Global theme data.

I'm also very concerned about introducing additional complexity being added to style different components when it's not needed. Patterns is being used as a synonym for variants. Which is confusing both terminology. Patterns typically in other design systems refers a way of combining multiple components together into a view to solve a common problem.

Examples: 
* https://www.lightningdesignsystem.com/guidelines/user-engagement/patterns/
* https://polaris.shopify.com/patterns/layout

https://github.com/square/maker/pull/294 and https://github.com/square/maker/pull/296 solve the problems in a similar way from this PR, they're non-breaking, and avoids skewing away from leveraging a global theme spec to support passing the required data for typography.

Please close this PR so we can move forward with a non-breaking solution and revisit the challenges with styling component variants later.

**Comment by pretzelhammer** [comment]
> This PR doesn't follow the consensus we had of following the Theme UI spec for the Global theme data.

It's not a consensus, I never agreed to it. I said in our meeting that we can use it for inspiration but we should never use it as an excuse to create friction and block work from going out. Right now it's being used as an excuse to create friction and block work from going out.

> I'm also very concerned about introducing additional complexity being added to style different components when it's not needed. 

It's needed. It was requested by Lauren and Bashu for Website work. Also patterns have already been merged and is in beta, that work isn't this PR, it doesn't make sense to block this PR on changes that aren't even in this PR. I also just had a convo with Shimaa and she confirmed that she thinks it's a great idea and we should add it to Maker. Engineering and Design are fully onboard with this going out.

> https://github.com/square/maker/pull/294 and https://github.com/square/maker/pull/296 solve the problems in a similar way from this PR

Those PRs are half-measures. This PR subsumes those PRs in functionality and implementation. Once we merge this PR there is no need for those PRs.

> they're non-breaking,

I already queried all of the engineers who would be impacted by the breaking changes and they're onboard with them. Design is also onboard.

> Please close this PR so we can move forward with a non-breaking solution and revisit the challenges with styling component variants later.

No. I want to deliver the work that engineers have asked for and that design has signed off on. That work includes fixes, features, and breaking changes. The breaking changes are not a downside but they are a benefit of this PR. The work is done. It has been reviewed. It is ready to merge. There's no reason to block this.

**Comment by landondurnan** [comment]
There is an alternative approach to supporting everything that's in the beta without adding the additional complexity of variants or patterns. Both of those concepts take us down a path we shouldn't be going down right now and makes it much harder to support the future goals we have for the system. We do not have consensus as a team on these concepts and they are not a priority for the current projects in flight. I am not okay with what's currently in the beta branch due to the increased complexity that can be completely avoided with simpler solutions.

We should prioritize based on the current requirements to support our goals for typography. https://github.com/square/maker/pull/294 and https://github.com/square/maker/pull/296 meets the immediate needs without the additional cost of breaking changes and complexity to the system.

We can immediately continue conversations around these concepts and alternative approaches when there's more time to do so.

**Comment by landondurnan** [comment]
> Is there a nomenclature that would make more sense to you @landondurnan?

Variants and passing data to component variants are something we are likely to support in the future, it's a common definition in all design systems and would be compatible with the Theme UI spec: https://theme-ui.com/theme-spec/#variants

However, the priority right now should be in getting the global data defined for colors, fonts, and upcoming with shape. After that I think we would benefit from exposing and reusing more CSS from that global theme data that's agnostic to the components. For our current existing design options that we support that would likely be everything we need for several months, but if not adding variant support to more components would make sense at that point.  But I think we should be considering everything a component variant can do to support advanced use cases and custom theming options beyond the design controls we currently support.

**Comment by landondurnan** [comment]
> Right now it feels like it's being used an excuse to block work from going out.

The complexity of completely new concepts that we haven't had time to do a complete deep dive on the everything we would need to support with these options is what I'm blocking right now.  It's not just not liking the name of something, although the name of things is important, it's considering the long-term impact and how we scale, test, and validate how this will be used in the future.  Does it solve a "now" problem? Sure. But it's a completely new way of doing something that can be done right now in a simpler, more-compliant, and easier solution.

**Comment by pretzelhammer** [comment]
@landondurnan 

This PR does not add patterns to Maker. Patterns are already in Maker on `beta`. If you have issues with patterns save that discussion for when a `beta` -> `master` PR is made. Nothing you have said is relevant to this PR. You're blocking this PR for changes that are not in this PR.

Since don't seem to have any actual objections to any of the actual content in this PR can you please approve it already. If you want to debate shipping user-defined patterns in Maker that discussion should be had in the `beta` -> `master` PR when that is made.

**Review by landondurnan — APPROVED** [review]
There's some things I think we should change with the theme data input, naming conventions, and how we support component variant styling, but since everything is in a beta these issues can be addressed in follow-up work with more testing.

**Pull request merged**
